spec/amp-framework-hosting.md

```diff
@@ -0,0 +1,274 @@
+
+- `--config`: Indicate the release type, production (`prod`) or canary (`canary`). Defaults to `prod`.
+- `--version_override`: Assign a version to the distribution. The version must consist of 13-digits. Defaults to the latest git commit time of the active branch.
+- `--sourcemap_url`: Provide the base URL for JavaScript source map links. This URL should contain placeholder `{version}` that will be replaced with the actual version when the AMP framework is built, for example `https://raw.githubusercontent.com/<github-username>/amphtml/{version}/`. Defaults to `http://localhost:8000/`.
+
+### Option 2: Download the framework with an AMP Toolbox tool
+
+[AMP Toolbox](https://github.com/ampproject/amp-toolbox) has both a Node.js module and a command line tool that will fetch a complete AMP framework from `cdn.ampproject.org`. Pick the tool best suited to your release workflow.
+
+- [@ampproject/toolbox-runtime-fetch](https://github.com/ampproject/amp-toolbox/tree/master/packages/runtime-fetch) - Node.js module
+- [@ampproject/toolbox-cli](https://github.com/ampproject/amp-toolbox/tree/master/packages/cli) - command line interface
+
+### Option 3: Manually copy the framework from cdn.ampproject.org
+
+The AMP framework can be copied from `cdn.ampproject.org`. The latest weekly release is always served from the root of `cdn.ampproject.org`. All [non-deprecated releases](https://github.com/ampproject/amphtml/blob/master/spec/amp-versioning-policy.md#version-deprecations) can be found in versioned URLs: `cdn.ampproject.org/rtv/<rtv>`, where `<rtv>` is the runtime version.
+
+Note: The AMP Project is looking into options for packaging releases ([#27726](https://github.com/ampproject/amphtml/issues/27726)).
+
+#### Copy files
+
+A listing of files in each release can be found in `files.txt` at the root of the framework distribution. For example, the files included in the current weekly release are listed in [cdn.ampproject.org/files.txt](https://cdn.ampproject.org/files.txt). Use your favorite HTTP client to download all of the files in `files.txt`, retaining path structures.
+
+#### Undo amp-geo dynamic modification
+
+When you request `amp-geo-0.1.js` from `cdn.ampproject.org` using an HTTP client, the CDN detects the country where the request originated and patches `amp-geo-0.1.js` on-the-fly. This patch needs to be reversed to ensure users are not all assigned the same country when amp-geo loads.
+
+When `cdn.ampproject.org` serves `amp-geo-0.1.js`, it replaces string `{{AMP_ISO_COUNTRY_HOTPATCH}}` with an ISO 3166-1 country code or an ISO 3166-2 country-subdivision code, followed by enough spaces to maintain the length of the string being replaced. Reversal of this patch can be accomplished by a RegEx replacement: search for `/[a-zA-Z]{2}(?:-[a-zA-Z0-9]{1,3} {22,24}| {26})/` and replace with `{{AMP_ISO_COUNTRY_HOTPATCH}}`.
```

Comment:
> it replaces string `{{AMP_ISO_COUNTRY_HOTPATCH}}` with an ISO 3166-1 country code or an ISO 3166-2 country-subdivision code

Technical nits: We don't insert a Country Code if we can't geolocate. If we can, then we insert it. And if we can geolocate, we'll _maybe_ insert a Subdivision, too. So, the format is either:

```
"{{AMP_ISO_COUNTRY_HOTPATCH}}" // should never happen from CDN
"                            " // 28 spaces
"CC                          " // 26 spaces
"CC CC-SD                    " // 20 spaces
```

So a regex might be something like `/ {28}|[A-Z]{2} {26}|[A-Z]{2} [A-Z]{2}-[A-Z]{2} {20}/i`

Reply:
Thanks for clarifying this. I didn't realize the ISO 3166-2 country-subdivision code is inserted independent of the country code. I just tested in California and I do see
```
"us us-ca                    "
```

I'm under the impression that broader subdivision support will be possible one day, so I'm leaning towards
```
/ {28}|[a-zA-Z]{2} {26}|[a-zA-Z]{2} [a-zA-Z]{2}-[a-zA-Z0-9]{1,3} {19,21}/
```

The `{1,3}` and `{19,21}` look a little fuzzy, but they greedy-match, so there wouldn't be any whitespace leftover with a code like `fj-w`. Note: I'm avoiding `/i` because some folks might use JavaScript.

Reply:
Pretty much all regex engines support the case-insensitive flag, and it really shrinks the regex itself and makes it a bit easier to read, so I think `//i` should be fine/clear enough.

Reply:
Done.